src/udev/udevadm-wait.c

```diff
@@ -410,16 +410,20 @@ int wait_main(int argc, char *argv[], void *userdata) {
         if (r < 0)
                 return log_error_errno(r, "Failed to set up inotify: %m");
 
-        r = setup_monitor(event, &monitor);
+        r = setup_monitor(event, MONITOR_GROUP_UDEV, "udev-uevent-monitor-event-source", &udev_monitor);
         if (r < 0)
-                return log_error_errno(r, "Failed to set up device monitor: %m");
+                return log_error_errno(r, "Failed to set up udev uevent monitor: %m");
 
         if (arg_wait_until == WAIT_UNTIL_ADDED) {
                 /* If --initialized=no is specified, it is not necessary to wait uevents for the specified
-                 * devices to be processed by udevd. Let's periodically check the devices. Then, we may be
-                 * able to finish this program earlier when udevd is very busy.
-                 *
-                 * This is useful for working around issues #24360 and #24450.
+                 * devices to be processed by udevd. Hence, let's listen on the kernel's uevent stream. Then,
+                 * we may be able to finish this program earlier when udevd is very busy.
+                 * Note, we still need to also setup udev monitor, as this may be invoked with a devlink. */
+                r = setup_monitor(event, MONITOR_GROUP_KERNEL, "kernel-uevent-monitor-event-source", &kernel_monitor);
```

Comment:
why listen to both? why not just the kernel group if we don't need udev, and not bother with udev?

Reply:
As explained in the comment, user may invoke with a devlink. In that case, we need to wait the uevent processed by udevd.

Reply:
ahh, thats a good point